src/fraction/approx.rs

```diff
@@ -0,0 +1,337 @@
+//! Approximate mathematical operations.
+//!
+//! This module implements operations that cannot guarantee lossless results, but which are
+//! nonetheless useful. Using any functionality from within this module requires a compromise to be
+//! made between performance and accuracy.
+
+use crate::{
+    convert::TryToConvertFrom, generic::GenericInteger, DynaFraction, DynaInt, GenericFraction,
+    Sign,
+};
+use num::{
+    bigint::ToBigInt, rational::Ratio, BigInt, BigRational, BigUint, Integer, ToPrimitive, Zero,
+};
+
+/// The level of accuracy to which a square root will be computed.
+#[derive(Clone, Debug)]
+pub struct SqrtAccuracy {
+    /// The value by which we multiply and divide in order to truncate values.
+    ///
+    /// Truncation is implemented here by shifting the decimal place by multiplying by some power
+    /// of 10, removing the fractional part of the result, then shifting the decimal place back by
+    /// dividing by the same power of 10. This value is that power of 10.
+    multiplier: BigInt,
+}
+
+impl SqrtAccuracy {
+    /// Creates a new `SqrtAccuracy` with `tail` digits after the decimal point.
+    pub fn new(tail: u32) -> SqrtAccuracy {
+        SqrtAccuracy {
+            multiplier: BigInt::from(10u8).pow(tail),
+        }
+    }
+
+    /// Truncates the ratio formed by `numer` and `denom` so that its decimal equivalent has only
+    /// the level of accuracy specified when `self` was created. This method takes ownership of
+    /// `numer` by necessity, but `denom` does not need to be mutated and thus only a reference to
+    /// it is required.
+    ///
+    /// This does **not** simplify the return value (hence `_raw`).
+    fn truncate_ratio_raw(&self, mut numer: BigInt, denom: &BigInt) -> BigRational {
+        numer *= &self.multiplier;
+
+        // Integer division gets rid of any digits we don't want.
+        numer /= denom;
+
+        // We now have an integer, so we can 'divide' by just giving a denominator other than 1 -
+        // in this case, we need to divide by the multiplier again.
+        Ratio::new_raw(numer, self.multiplier.clone())
+    }
+}
+
+/// An approximation of a square root.
+pub enum SqrtApprox {
+    /// A rational (i.e. fractional) approximation.
+    ///
+    /// Depending on the accuracy used, these numbers can be *very* large to store (>100 KB with
+    /// excessive accuracy), so cloning is likely to be expensive. This size also means that
+    /// computations can take a long time.
+    Rational(BigRational),
```

Comment:
I had done this originally, but I switched to `BigRational` because it's easier to work with (it has `from_float`, for example). I've just switched it all back and it wasn't as difficult as I thought, so thanks for the suggestion.